### PR TITLE
PLT-97 Add style guide and a link to it from CONTRIBUTING doc.

### PR DESCRIPTION
**Pull request by thealmarty**
- I commented out a couple things that we aren't very sure about. Changed the wording a bit in a couple places. 
- Added a link under the "Working conventions" section, which maybe could use a tidy up now that we have the style guide. It's also OK as is, it's not very repetitive. 

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: CONTRIBUTING.adoc [hunk 1]
```diff
@@ -167,6 +167,10 @@ You can update it xref:update-nix-pins[with the Nix flake commands].
 
 == Working conventions
 
+=== Style guide
+
+Please follow our link:STYLEGUIDE{outfilesuffix}[Haskell style guide], which documents most of our conventions.
```
Comment (michaelpj) [inline CONTRIBUTING.adoc:172]:
```suggestion
Please follow our link:STYLEGUIDE{outfilesuffix}[Haskell style guide], which documents most of our conventions for working on Haskell code.
```

File: STYLEGUIDE.adoc [hunk 2]
```diff
@@ -0,0 +1,426 @@
+= Plutus Haskell style guide
+
+The purpose of this document is to codify various aspects of how we write Haskell code as we contribute to Plutus. 
+
+The goal is twofold:
+To provide guidance in cases where it is not obvious what to do, or the choice is arbitrary.
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:6]: bullet points got lost?

File: STYLEGUIDE.adoc [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+
+The contents of this document are “recommendations”, that is:
+They are more concrete than “principles” (like “write simple code”) because we want them to be clearly action-guiding.
+They are more vague than “rules” (like “never use function X”) because it’s usually impossible to be that precise, and so judgement is still required in applying them.
+
+== Haskell
+
+=== Basic formatting
+
+Use 4 spaces for indentation. Avoid lines over 120 characters. (This is codified in `.editorconfig`.)
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:17]: these parens are silly actually
```suggestion
Use 4 spaces for indentation. Avoid lines over 120 characters. This is codified in `.editorconfig`.
```

File: STYLEGUIDE.adoc [hunk 4]
```diff
@@ -0,0 +1,426 @@
+= Plutus Haskell style guide
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:1]: add a TOC with `:toc:`?

File: STYLEGUIDE.adoc [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+Every top-level binding should have a type signature. _Most_ other named bindings (e.g. `let` and `where` bindings) should also have type signatures.
+
+==== Reasoning
+
+Haskell is traditionally lauded for its excellent type inference. People used to make the argument that this brought Haskell closer to dynamically-typed languages in ease of use: you don't have to write type signatures, the compiler will infer it for you.
+
+However, not writing type signatures has a heavy _maintainability_ cost.
+
+.1. Non-local errors
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:39]: This seems to format a bit weirdly. Probably these should just be lower-level headings?

File: STYLEGUIDE.adoc [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+
+Changing the type of a binding often means that something relatively significant has changed. But if the type is inferred, this can happen without you noticing it. This is almost always bad!
+
+=== Write kind signatures
+
+Every type definition that has parameters which are not all of kind `Type` should have a kind signature using `StandaloneKindSignatures`.
+
+==== Examples
+Bad:
+```
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:61]: Asciidoc has source blocks with specific languages, which in principle should get us syntax higlighting: https://docs.asciidoctor.org/asciidoc/latest/verbatim/source-blocks/

File: STYLEGUIDE.adoc [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+Good:
+```
+for xs $ \x -> …
+```
+
+==== Reasoning
+
+Haskell can get quite condensed and hard to read, especially when the reading direction changes frequently. Often there are symmetrical versions of operators like `<=<` and `>=>` or `=<<` and `>>=` that you can switch between to make code easier to read.
+=== When to use hierarchical re-exporting modules
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:231]:
```suggestion

=== When to use hierarchical re-exporting modules
```

File: STYLEGUIDE.adoc [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+
+TODO
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:233]: I don't generally think it's worth including such TODOs, I'm not sure it makes us any more likely to remember to write something.

File: STYLEGUIDE.adoc [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+
+See the blog post.
+
+=== Extensions
+
+==== The Good
+
+These are basically all fine and can be put in `default-extensions`.
+
+1. Anything in https://github.com/ghc-proposals/ghc-proposals/pull/380[`GHC2021`]. Once we have a GHC version that supports the GHC2021 language, we will likely switch to using it.
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:249]: semantically these seem like unordered lists, but not a big deal

File: STYLEGUIDE.adoc [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+Good:
+```
+import Control.Lens qualified as Lens
+```
+
+==== Reasoning
+
+Complex import statements are difficult to maintain and cause annoying diffs which are also hard to merge. Qualified function usages are quite easy to read, and not that much worse to write.
+
+// Group your imports by source
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:298]: just delete it

File: STYLEGUIDE.adoc [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+
+==== Reasoning
+
+`Default` is not terribly bad. It’s truly ad-hoc polymorphism: all you get is name reuse, you can’t (or shouldn’t) write a function that’s polymorphic over `Default a`. That’s fine, but it also means that the benefit is fairly marginal.
+
+Additionally, just defining specific values is more flexible. If, say, you want multiple default values for different contexts, then that is straightforward with values but not with `Default`.
+
+== Cabal
+
+// === Metadata
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:354]: delete

File: STYLEGUIDE.adoc [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+
+It’s nice for files to be self-contained, but this is typically a fiction: you need to know about compilation flags from cabal files anyway. It saves a lot of typing to put the really essential stuff in the cabal file.
+
+Haskell “languages” are basically a blessed set of extensions anyway, and people are fine putting those in the cabal file. A lot of what we’re currently doing is manually implementing the GHC2021 language!
+
+=== Warning flags
+
+Use the following set of warning flags:
+
+-Wall
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:379]: format with some verbatim formatting

**Review by michaelpj — COMMENTED** [review]
A few comments. Worth looking at the rich diff on Github, makes rendering issues more obvious!

**Comment by thealmarty** [comment]
> A few comments. Worth looking at the rich diff on Github, makes rendering issues more obvious!

Sorry I was using the vscode preview function, maybe that doesn't render the same as on here. I'll remember that for next time.

**Comment by michaelpj** [comment]
> Sorry I was using the vscode preview function, maybe that doesn't render the same as on here. I'll remember that for next time.

Huh, weird? I think asciidoc isn't fully standardized, so maybe that's it...

File: STYLEGUIDE.adoc (hunk 5 above)
Comment (thealmarty) [reply STYLEGUIDE.adoc:39]: Actually this is rendering right on github it seems (just like in the google doc). But yeah I can make them headings instead.

Comment (michaelpj) [reply STYLEGUIDE.adoc:39]: I get strange stuff where it makes weird collapsible sections or something?

**Comment by thealmarty** [comment]
I think I got everything? Ready to merge once approved.

File: STYLEGUIDE.adoc [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+
+It’s nice for files to be self-contained, but this is typically a fiction: you need to know about compilation flags from cabal files anyway. It saves a lot of typing to put the really essential stuff in the cabal file.
+
+Haskell “languages” are basically a blessed set of extensions anyway, and people are fine putting those in the cabal file. A lot of what we’re currently doing is manually implementing the GHC2021 language!
+
+=== Warning flags
+
+Use the following set of warning flags:
+
+[source,Haskell]
```
Comment (michaelpj) [inline STYLEGUIDE.adoc:418]: These aren't haskell! So it highlights weirdly.

**Review by michaelpj — APPROVED** [review]
Just one tiny thing, then feel free to merge.

**Pull request merged**